eclipse-theia/theia
# Implement Tree Compression

#### What it does


Closes #4783
Closes #10268 

This implements compact nodes for `TreeNode`s with a single child that is also a `CompositeTreeNode` and uses those services in the `NavigatorTree`. It is based mostly on #10268.

I also refactored the `createTreeContainer` method to reduce the amount of boilerplate you have to write to set up a custom tree. It's not really necessary, but it does both make it clearer to callers what they get in their container (interfaces define the full service set) and make it easier to override those services.

#### How to test


1. This commit includes a demo workspace under `api-samples` that has various structures of interest to compression. Open Theia or open that workspace directly.
2. Test various behaviors with compressed nodes:
 - keyboard navigation (up-down should always move up or down rows, right-left should expand / collapse only at the ends of rows)
 - click selection / expansion / collapse
 - drag-drop (beware of #10712)
3. Test changing the `explorer.compactFolders` preference and observe that the tree compresses and uncompresses as expected.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
